src/Data/List/Sort/InsertionSort.agda

```diff
@@ -0,0 +1,230 @@
+
+module Data.List.Sort.InsertionSort
+  {a ℓ₁ ℓ₂}
+  (O : DecTotalOrder a ℓ₁ ℓ₂)
+  where
+
+open import Data.Bool.Base using (true; false ; if_then_else_)
+open import Data.Empty using (⊥-elim)
+open import Function using (id)
+
+open import Data.List.Base using (List; []; _∷_)
+open import Data.List.Relation.Unary.Linked using ([]; [-] ; _∷_)
+open import Data.List.Relation.Binary.Pointwise
+  using (Pointwise ; [] ; _∷_ ; decidable ; setoid)
+
+open import Relation.Binary.Bundles using (Setoid)
+open import Relation.Binary.Definitions using (Decidable)
+open import Relation.Binary.PropositionalEquality
+  using (inspect ; [_])
+open import Relation.Binary.Properties.DecTotalOrder O
+  using (≰⇒≥ ; ≥-antisym)
+open import Relation.Nullary.Decidable.Core
+  using (does ; yes ; no ; map′)
+
+open DecTotalOrder O
+  renaming (Carrier to A ; refl to reflA ; trans to transA)
+  using (totalOrder ; _≤?_ ; _≤_ ;
+  module Eq ; _≈_ ; ≤-respʳ-≈ ; ≤-respˡ-≈ ; antisym)
+  
+open import Data.List.Relation.Binary.Equality.Setoid Eq.setoid
+  using (_≋_)
+open import Data.List.Relation.Binary.Permutation.Setoid Eq.setoid
+open import Data.List.Relation.Unary.Sorted.TotalOrder totalOrder
+  using (Sorted)
+open import Data.List.Sort.Base totalOrder
+open import Relation.Binary.Reasoning.Setoid (setoid Eq.setoid)
+  as SetoidReasoning using ()
+
+open Setoid (setoid Eq.setoid)
+  renaming (refl to reflₚₜ ; trans to transₚₜ) using ()
+
+------------------------------------------------------------------------
+-- Definitions
+
+insert : A → List A → List A
+insert x [] = x ∷ []
+insert x (y ∷ xs) = if does (x ≤? y)
+  then x ∷ y ∷ xs
+  else y ∷ insert x xs
+
+sort : List A → List A
+sort [] = []
+sort (x ∷ xs) = insert x (sort xs)
+
+------------------------------------------------------------------------
+-- Permutation property
+
+insert-↭ : ∀ x xs → insert x xs ↭ x ∷ xs
+insert-↭ x [] = ↭-refl
+insert-↭ x (y ∷ xs) with does (x ≤? y)
+... | true = ↭-refl
+... | false = begin
+  y ∷ insert x xs ↭⟨ prep Eq.refl (insert-↭ x xs) ⟩
+  y ∷ x ∷ xs      ↭⟨ swap Eq.refl Eq.refl ↭-refl ⟩
+  x ∷ y ∷ xs ∎
+  where open PermutationReasoning
+
+insert-cong-↭ : ∀ {x xs x' xs'} → x ≈ x' → xs ↭ xs' →
+                insert x xs ↭ x' ∷ xs'
+insert-cong-↭ {x} {xs} {x'} {xs'} eq1 eq2 = begin
+  insert x xs ↭⟨ insert-↭ x xs ⟩
+  x ∷ xs      ↭⟨ prep eq1 eq2 ⟩
+  x' ∷ xs' ∎
+  where open PermutationReasoning
+
+sort-↭ : ∀ (xs : List A) → sort xs ↭ xs
+sort-↭ [] = ↭-refl
+sort-↭ (x ∷ xs) = insert-cong-↭ Eq.refl (sort-↭ xs)
+
+------------------------------------------------------------------------
+-- Sorted property
+
+insert-↗ : ∀ x {xs} → Sorted xs → Sorted (insert x xs)
+insert-↗ x [] = [-]
+insert-↗ x ([-] {y}) with (x ≤? y)
+... | yes x≤y = x≤y ∷ [-]
+... | no x≤y = ≰⇒≥ x≤y ∷ [-]
+insert-↗ x (_∷_ {y} {z} {ys} y≤z z≤ys) with (x ≤? y)
+... | yes x≤y = x≤y ∷ y≤z ∷ z≤ys
+... | no x≤y with insert-↗ x z≤ys
+... | sd with (x ≤? z)
+... | yes r = ≰⇒≥ x≤y ∷ sd
+... | no r = y≤z ∷ sd
```

Comment:
This can be streamlined to
```suggestion
... | no x≤y with sd ← insert-↗ x z≤ys | x ≤? z
...   | yes r = ≰⇒≥ x≤y ∷ sd
...   | no r = y≤z ∷ sd
```
(and where, again, the `r`s are redundant, so could be replaced by `_`)